File: D.S.A_python/binary_search.py
#binary search
import time
import random

t = int(input("enter the no. of test cases: "))

def binarysearch(arr, n, x):
    low = 0
    high = n - 1

    while low <= high:
        mid = (low + high) // 2

        if arr[mid] == x:
            return mid
        elif arr[mid] < x:
            low = mid + 1
        else:
            high = mid - 1

    return -1


for _ in range(t):
    start = time.perf_counter()

    n = int(input("enter the size: "))
    arr = [random.randint(-10,10) for _ in range(n)]

    arr.sort() 

    for i in range(n):
        print(arr[i], end=" ")
    print()

    x = int(input("enter the element you want: "))

    result = binarysearch(arr, n, x)

    if result != -1:
        print("the element found at the index:", result)
    else:
        print("element not found!!")

    end = time.perf_counter()
    print(f"time = {end - start}")

File: D.S.A_c/binary_search.c
#include<stdio.h>
int binary(int arr[],int size,int search){
    int left=0,right=size-1;int mid;
    while(left<=right){
        mid=left+(right-left)/2;
        if(arr[mid]==search)
        return mid;

        else if(arr[mid]<search)
        left = mid + 1;

        else if(arr[mid]>search)
        right = mid - 1;
    }
    return -1;
}

int main(){
    int n;int search;
    scanf("%d",&n);
    int array[n];
    for(int i=0;i<n;i++){
        scanf("%d",&array[i]);
    }
    printf("Enter element to search: ");
    scanf("%d",&search);

    int index=binary(array,n,search);
    if(index!=-1)
    printf("Element found at index %d",index);
    else
    printf("Element not found!");
    return 0;
}